src/core/queryObserver.ts

```diff
@@ -148,6 +148,37 @@ export class QueryObserver<
     this.currentQuery.removeObserver(this)
   }
 
+  createQueryResult(
+    queryObserverResult: QueryObserverResult<TData, TError>
+  ): QueryObserverResult<TData, TError> {
+    if (this.options.notifyOnChangeTracked) {
+      const trackedResult = {}
+      const addNotifyOnChangeProps = (prop: keyof QueryObserverResult) => {
+        if (!this.options.notifyOnChangeProps) {
+          this.options.notifyOnChangeProps = []
+        }
+
+        if (!this.options.notifyOnChangeProps.includes(prop)) {
+          this.options.notifyOnChangeProps.push(prop)
```

Comment:
Not sure if it is a good idea to mutate the `notifyOnChangeProps` option. What about storing the accessed properties in a separate array and then merge it with `notifyOnChangeProps` in `shouldNotifyListeners`?

Reply:
did that here: ba1bb199

I think code will be a bit easier if we decide to go with this proposal though: https://github.com/tannerlinsley/react-query/pull/1578#discussion_r554466089